### PR TITLE
`delete_network_options` should only set network `notoptions` on multisites

### PR DESCRIPTION
**Pull request by anomiex**
On a non-multisite, `get_network_option()`, `add_network_option()`, and `update_network_option()` don't bother updating the network `notoptions` cache, instead reling on `get_option()`, `add_option()`, and `update_option()` to use the non-network `notoptions`.

`delete_options()` and `delete_network_options()` were recently updated to populate the `notoptions` caches. But, contrary to existing practice, `delete_network_options()` was made to update the network `notoptions` cache even on non-multisites.

This moves the updating of the network `notoptions` cache in `delete_network_options()` to the multisite code path, matching the behavior of the other functions, and adds a test to verify this.

Trac ticket: https://core.trac.wordpress.org/ticket/61730

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by anomiex** [comment]
An alternative fix would be to update the other functions to all populate the network `notoptions` cache, but this seemed more straightforward.

**Review by peterwilsoncc — APPROVED** [review]
This looks good to me, thank you!

I created https://github.com/WordPress/wordpress-develop/pull/7079 to:

* ensure the new test added here failed without the fix
* add some additional tests
* push the code change to ensure the tests started passing

It all worked as expected so I have taken the liberty of pushing 3ec1a4d81db74e7293cf65873abe8d6b5352408d to this branch with the additional tests. This will allow us to work from a single PR.

@pbearne are you able to take a look at the updated code?

File: src/wp-includes/option.php [hunk 1]
```diff
@@ -2264,6 +2264,17 @@ function delete_network_option( $network_id, $option ) {
 				'site_id'  => $network_id,
 			)
 		);
+
+		if ( $result ) {
```
Comment (peterwilsoncc) [inline src/wp-includes/option.php:2268]: Sorry, on second thoughts I am wondering if this should be 

```suggestion
		if ( false !== $result ) {
```

If `$wpdb::delete()` returns `0`, it means that the query ran successfully but there was no option to delete. The result is the same, it's a known notoption.

If `$wpdb::delete()` returns `false`, then the query failed so it's not known whether or not the option doesn't exist.

Comment (anomiex) [reply src/wp-includes/option.php:2268]: If there's a problem with the code here, there's the same problem already in trunk: the notoptions-setting code on [lines 2296–2303](https://github.com/WordPress/wordpress-develop/blob/5a76260c3cd22d4688b91620ad7a00a42300c707/src/wp-includes/option.php#L2296-L2303) is guarded by an `if ( $result )` on [line 2269](https://github.com/WordPress/wordpress-develop/blob/5a76260c3cd22d4688b91620ad7a00a42300c707/src/wp-includes/option.php#L2269).

`delete_option()` then has the opposite problem, it sets `notoptions` even if `$result === false`.

Comment (peterwilsoncc) [reply src/wp-includes/option.php:2268]: > `delete_option()` then has the opposite problem, it sets `notoptions` even if `$result === false`.

Good catch. I'll open the original ticket to note that it needs a little work to account for a database failure.

Are you happy to add some protections here or would you prefer me to do it on a separate PR?

Comment (anomiex) [reply src/wp-includes/option.php:2268]: I think I'd prefer you fix these issues in a separate PR.

**Comment by peterwilsoncc** [comment]
Merged [r58811](https://core.trac.wordpress.org/changeset/58811) / https://github.com/WordPress/wordpress-develop/commit/b38541de39c5e3ccb75ff76b49ab930846c878b6

Thanks for catching this, Brad, it's much appreciated.

**Pull request closed without merging**